sodium/src/pwhash.rs

```diff
@@ -0,0 +1,145 @@
+//! This module provides access to libsodium
+
+use super::{check_init, secbuf::SecBuf};
+use crate::{error::SodiumResult, random::random_secbuf};
+
+pub const OPSLIMIT_INTERACTIVE: u64 = rust_sodium_sys::crypto_pwhash_OPSLIMIT_INTERACTIVE as u64;
+pub const MEMLIMIT_INTERACTIVE: usize =
+    rust_sodium_sys::crypto_pwhash_MEMLIMIT_INTERACTIVE as usize;
+pub const OPSLIMIT_MODERATE: u64 = rust_sodium_sys::crypto_pwhash_OPSLIMIT_MODERATE as u64;
+pub const MEMLIMIT_MODERATE: usize = rust_sodium_sys::crypto_pwhash_MEMLIMIT_MODERATE as usize;
+pub const OPSLIMIT_SENSITIVE: u64 = rust_sodium_sys::crypto_pwhash_OPSLIMIT_SENSITIVE as u64;
+pub const MEMLIMIT_SENSITIVE: usize = rust_sodium_sys::crypto_pwhash_MEMLIMIT_SENSITIVE as usize;
+
+pub const ALG_ARGON2I13: i8 = rust_sodium_sys::crypto_pwhash_ALG_ARGON2I13 as i8;
+pub const ALG_ARGON2ID13: i8 = rust_sodium_sys::crypto_pwhash_ALG_ARGON2ID13 as i8;
+
+pub const HASHBYTES: usize = 32 as usize;
+pub const SALTBYTES: usize = rust_sodium_sys::crypto_pwhash_SALTBYTES as usize;
+
+/// Calculate a password hash
+///
+/// @param {SecBuf} password - the password to hash
+///
+/// @param {u64} opslimit - operation scaling for hashing algorithm
+///
+/// @param {usize} memlimit - memory scaling for hashing algorithm
+///
+/// @param {i8} algorithm - which hashing algorithm
+///
+/// @param {SecBuf} salt - optional predefined salt (random if not included)
+///
+/// @param {SecBuf} hash - the hash generated
+pub fn hash(
+    password: &mut SecBuf,
+    ops_limit: u64,
+    mem_limit: usize,
+    alg: i8,
+    salt: Option<&mut SecBuf>,
+    hash: &mut SecBuf,
+) -> SodiumResult<()> {
+    check_init();
+    let my_salt_locker;
+    let mut random_salt = SecBuf::with_insecure(SALTBYTES);
+    random_secbuf(&mut random_salt);
+    let random_salt = random_salt.read_lock();
+    let mut my_salt = raw_ptr_char_immut!(random_salt);
+
+    if let Some(s) = salt {
+        my_salt_locker = s.read_lock();
+        my_salt = raw_ptr_char_immut!(my_salt_locker);
+    }
```

Comment:
I think we need to take the `Option<>` off the salt... If we are using a random salt, we need to return it to the user somehow, or they'll never be able to repeat the hashing. API-wise it will be simpler to always take a salt, and if the user wants a random one, they can radomize it themselves.